### PR TITLE
(MQ cleanup) Use BinaryPrimitives where possible

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
This finds places in the code where we're reading data one byte at a time to build up a larger integer, and it replaces those patterns with calls to `BinaryPrimitives` instead.

Since `BinaryPrimitives` wasn't introduced until netcoreapp2.1, I didn't touch projects like _System.Diagnostics.EventLog_ that target earlier runtimes. I also didn't touch existing netcoreapp2.1 projects if it would have meant adding a _System.Memory_ package dependency where such direct dependency didn't previously exist. I also didn't touch the wasm-specific crypto code which was essentially a direct port from .NET Framework 4.x. That code's being deleted eventually anyway.

### Risks:

- I may have written `BigEndian` where I meant to write `LittleEndian` (or vice versa).
- I may have passed the wrong parameters to `MemoryExtensions.Slice`.
- If a previous data access would have failed due to an out-of-bounds exception, the original code path would have resulted in a `NullReferenceException` or an `IndexOutOfRangeException`. The new code paths instead will result in an `ArgumentOutOfRangeException`. I don't think anybody is relying on the old behavior because I believe every call site I touched is guarded with an explicit bounds check, but needs to be called out nonetheless.

### CODE REVIEW

**Comment by stephentoub** [comment]
This seems like something that could be automated to catch the majority of cases. Did you consider an analyzer/fixer?

**Comment by GrabYourPitchforks** [comment]
> Did you consider an analyzer/fixer?

Only briefly. The big marks against spending time working on an analyzer are that: (a) there didn't seem to be an egregiously large number of occurrences; and (b) even within this PR there were a few different variations on this pattern.

File: src/libraries/Common/src/System/Net/SocketAddressPal.Windows.cs [hunk 1]
```diff
@@ -26,80 +27,44 @@ public static unsafe void SetAddressFamily(byte[] buffer, AddressFamily family)
...
         {
             // IPv4 Address serialization
-            buffer[4] = unchecked((byte)(address));
-            buffer[5] = unchecked((byte)(address >> 8));
-            buffer[6] = unchecked((byte)(address >> 16));
-            buffer[7] = unchecked((byte)(address >> 24));
+            BinaryPrimitives.WriteUInt32LittleEndian(buffer.AsSpan(4), address);
         }
 
         public static unsafe void SetIPv6Address(byte[] buffer, Span<byte> address, uint scope)
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/SocketAddressPal.Windows.cs:58]: Can some of these `unsafe`s be removed?

File: src/libraries/Common/src/System/Net/SocketAddressPal.Windows.cs [hunk 2]
```diff
@@ -26,80 +27,44 @@ public static unsafe void SetAddressFamily(byte[] buffer, AddressFamily family)
...
         {
             // No handling for Flow Information
-            buffer[4] = (byte)0;
-            buffer[5] = (byte)0;
-            buffer[6] = (byte)0;
-            buffer[7] = (byte)0;
+            BinaryPrimitives.WriteUInt32LittleEndian(buffer.AsSpan(4), 0);
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/SocketAddressPal.Windows.cs:61]: Endianness here wouldn't matter.  But I guess it doesn't really matter perf-wise, especially on a little-endian arch.

File: src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs [hunk 3]
```diff
@@ -185,7 +185,7 @@ public IPAddress(ReadOnlySpan<byte> address)
         {
             if (address.Length == IPAddressParserStatics.IPv4AddressBytes)
             {
-                PrivateAddress = (uint)((address[3] << 24 | address[2] << 16 | address[1] << 8 | address[0]) & 0x0FFFFFFFF);
```
Comment (stephentoub) [inline src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs:188]: What a strange `0x0FFFFFFFF` masking.  I have to wonder what the original author was thinking there.

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs [hunk 4]
```diff
@@ -172,12 +174,15 @@ private unsafe void Dispose(bool disposing)
...
         {
-            byte* buffer = (byte*)p;
             // Unaligned, little endian format
-            return buffer[0] | (buffer[1] << 8) | (buffer[2] << 16) | (buffer[3] << 24);
+            int value = Unsafe.ReadUnaligned<int>(p);
+            if (!BitConverter.IsLittleEndian)
+            {
+                value = BinaryPrimitives.ReverseEndianness(value);
+            }
+            return value;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs:183]: Is this better than:
```C#
return BinaryPrimitives.ReadInt32LittleEndian(new Span<int>(p, sizeof(int)));
```
?

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs [hunk 5]
```diff
@@ -1433,25 +1434,13 @@ private static string TZif_GetZoneAbbreviation(string zoneAbbreviations, int ind
...
-        {
-            fixed (byte* pbyte = &value[startIndex])
-            {
-                int i1 = (*pbyte << 24) | (*(pbyte + 1) << 16) | (*(pbyte + 2) << 8) | (*(pbyte + 3));
-                int i2 = (*(pbyte + 4) << 24) | (*(pbyte + 5) << 16) | (*(pbyte + 6) << 8) | (*(pbyte + 7));
-                return (uint)i2 | ((long)i1 << 32);
-            }
-        }
+        private static long TZif_ToInt64(byte[] value, int startIndex)
+            => BinaryPrimitives.ReadInt64BigEndian(value.AsSpan(startIndex));
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs:1443]: It looks like both of these methods are used in just one location below.  Maybe both of these should be deleted and then TZif_ToUnixTime rewritten as:
```C#
private static long TZif_ToUnixTime(byte[] value, int startIndex, TZVersion version)
{
    Span<byte> span = value.AsSpan(startIndex);
    return version != TZVersion.V1 ?
        BinaryPrimitives.ReadInt64BigEndian(span) :
        BinaryPrimitives.ReadInt32BigEndian(span);
```
?

**Review by stephentoub — APPROVED** [review]
Nice. Thanks.

**Comment by GrabYourPitchforks** [comment]
Oh, the joys of not having a working build environment and relying on CI. ;)

**Comment by GrabYourPitchforks** [comment]
Failing OSX leg is known issue https://github.com/dotnet/runtime/issues/44037.

**Pull request merged**